resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/service/ResourceServicePersistenceProxy.java

```diff
@@ -0,0 +1,181 @@
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import uk.gov.gchq.palisade.resource.LeafResource;
+import uk.gov.gchq.palisade.service.ResourceService;
+import uk.gov.gchq.palisade.service.resource.exception.NoSuchResourceException;
+import uk.gov.gchq.palisade.service.resource.model.AuditErrorMessage;
+import uk.gov.gchq.palisade.service.resource.model.AuditableResourceResponse;
+import uk.gov.gchq.palisade.service.resource.model.ExceptionSource;
+import uk.gov.gchq.palisade.service.resource.model.ResourceRequest;
+import uk.gov.gchq.palisade.service.resource.model.ResourceResponse;
+import uk.gov.gchq.palisade.service.resource.repository.PersistenceLayer;
+
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * A proxy of (wrapper around) an instance of a {@link ResourceService}.
+ * This adds a cache-like behaviour to the service by persisting requests/responses in a database.
+ * Additionally, this is expected to be used by an asynchronous REST streaming response, so has further considerations
+ * to properly support the callback.
+ */
+public class ResourceServicePersistenceProxy {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceServicePersistenceProxy.class);
+    private static final String STORE = "Trying from persistence store";
+
+    private final PersistenceLayer persistence;
+    private final ResourceService delegate;
+
+    /**
+     * Construct a StreamingResourceServiceProxy, but without any {@link uk.gov.gchq.palisade.service.ResourcePrepopulationFactory} prepopulation
+     *
+     * @param persistence a {@link PersistenceLayer} for persisting resources in, as if it were a cache
+     * @param delegate    a 'real' {@link ResourceService} to delegate requests to when not found in the persistence layer
+     */
+
+    public ResourceServicePersistenceProxy(final PersistenceLayer persistence, final ResourceService delegate) {
+        this.persistence = persistence;
+        this.delegate = delegate;
+    }
+
+    /**
+     * Uses a resource to get any {@link LeafResource}s associated with the it.
+     *
+     * @param request the the {@link ResourceRequest} that contains the resourceId used to retrieve resources
+     * @return a {@link Source} of {@link LeafResource}s associated with the resource
+     */
+    public Source<AuditableResourceResponse, NotUsed> getResourcesByResource(final ResourceRequest request) {
+        return this.getResourcesById(request);
+    }
+
+    /**
+     * Uses a resourceId to get any {@link LeafResource}s associated with the it.
+     *
+     * @param request the the {@link ResourceRequest} that contains the resourceId used to retrieve resources
+     * @return a {@link Source} of {@link LeafResource}s associated with the resourceId
+     */
+    public Source<AuditableResourceResponse, NotUsed> getResourcesById(final ResourceRequest request) {
```

Comment:
Naming a little confusing here maybe
```suggestion
    public Source<AuditableResourceResponse, NotUsed> getResourcesByRequest(final ResourceRequest request) {
```

Reply:
I'm not sure this name change is necessary. We are getting the resources by id but we use the request to get that information and then create the response.